File: JavaScript/AoC2025/days/day9.js
import fs, { rmSync } from "fs"
import path from "path"

export default class Day9 {
    constructor() {
        const filePath = path.resolve("../../Data/2025/day9.txt");
        const data = fs.readFileSync(filePath, "utf-8");
        this.positions = data
            .split('\n')
            .map(line => line.trim())
            .filter(line => line.length > 0)
            .map(line => line.split(',').map(Number));
        this.rectangles = this.getOrderedRectangles();

        this.xCoords = this.getOrderedXCoordinates();
        this.yCoords = this.getOrderedYCoordinates();
        this.xMap = Object.fromEntries(this.xCoords.map((x, i) => [x, i]));
        this.yMap = Object.fromEntries(this.yCoords.map((y, i) => [y, i]));
        this.compressed = this.positions.map(([x, y]) => [this.xMap[x], this.yMap[y]]);

        this.edge = this.scanEdge();
    }

    printResults() {
        console.log("Day 9:");
        console.log(`Part 1: ${this.part1()}`);
        console.log(`Part 2: ${this.part2()}`);
    }

    part1() {
        return this.rectangles[0][0];
    }

    part2() {
        for (const rectangle of this.rectangles) {
            const area = rectangle[0];
            const a = this.compressed[rectangle[1]];
            const b = this.compressed[rectangle[2]];

            if (this.checkRectangle(a, b))
                return area;
        }

        return 0;
    }

    findArea(a, b) {
        return (Math.abs(a[0] - b[0]) + 1) * (Math.abs(a[1] - b[1]) + 1);
    }

    getOrderedRectangles() {
        const length = this.positions.length;
        const rectangles = [];
        for (let i = 0; i < length - 1; i++) {
            for (let j = i + 1; j < length; j++) {
                const area = this.findArea(this.positions[i], this.positions[j]);
                rectangles.push([area, i, j]);
            }
        }
        rectangles.sort((a, b) => b[0] - a[0]);
        return rectangles;
    }

    getOrderedXCoordinates() {
        const xCoords = this.positions.map(coord => coord[0]);
        xCoords.sort((a, b) => a - b);
        return xCoords;
    }

    getOrderedYCoordinates() {
        const yCoords = this.positions.map(coord => coord[1]);
        yCoords.sort((a, b) => a - b);
        return yCoords;
    }

    scanEdge() {
        const edge = new Set();
        const coords = this.compressed.slice();
        let prev = coords[0];
        coords.push(prev);
        for (const pos of coords.slice(1)) {
            const x1 = prev[0];
            const y1 = prev[1];
            const x2 = pos[0];
            const y2 = pos[1];

            if (x1 === x2) {
                for (let i = Math.min(y1, y2); i <= Math.max(y1, y2); i++) {
                    edge.add(`${x1},${i}`);
                }
            } else {
                for (let j = Math.min(x1, x2); j <= Math.max(x1, x2); j++) {
                    edge.add(`${j},${y1}`);
                }
            }

            prev = pos;
        }

        return edge;
    }

    checkRectangle(a, b) {
        const x1 = a[0];
        const y1 = a[1];
        const x2 = b[0];
        const y2 = b[1];

        const maxX = Math.max(x1, x2);
        const minX = Math.min(x1, x2);
        const maxY = Math.max(y1, y2);
        const minY = Math.min(y1, y2);

        for (let x = minX + 1; x < maxX; x++) {
            if (this.edge.has(`${x},${minY + 1}`) || this.edge.has(`${x},${maxY - 1}`))
                return false;
        }

        for (let y = minY + 1; y < maxY; y++) {
            if (this.edge.has(`${minX + 1},${y}`) || this.edge.has(`${maxX - 1},${y}`))
                return false;
        }

        return true;
    }
}

File: Python/AoC2025/days/day9.py
class Day9:
    def __init__(self):
        with open("../../Data/2025/day9.txt") as f:
            self.positions = [[int(x) for x in line.rstrip("\n").split(",")] for line in f]

        self.x_coords = sorted({x for x, y in self.positions})
        self.y_coords = sorted({y for x, y in self.positions})
        self.x_map = {x: i for i, x in enumerate(self.x_coords)}
        self.y_map = {y: i for i, y in enumerate(self.y_coords)}

        self.compressed = [(self.x_map[x], self.y_map[y]) for x, y in self.positions]

        self.rectangles = self.find_sorted_rectangles()
        self.edge = self.scan_edges()

    def part_one(self):
        return self.rectangles[0][0]

    def part_two(self):
        for area, i, j in self.rectangles:
            a = self.compressed[i]
            b = self.compressed[j]
            if self.check_rectangle(a, b):
                return area
        return 0

    def find_area(self, a, b):
        return (abs(a[0] - b[0]) + 1) * (abs(a[1] - b[1]) + 1)

    def find_sorted_rectangles(self):
        rectangles = []
        n = len(self.positions)
        for i in range(n - 1):
            for j in range(i + 1, n):
                area = self.find_area(self.positions[i], self.positions[j])
                rectangles.append([area, i, j])
        rectangles.sort(key=lambda x: x[0], reverse=True)
        return rectangles

    def scan_edges(self):
        edge = set()
        coords = self.compressed.copy()
        prev = coords[0]
        coords.append(prev)
        for curr in coords[1:]:
            x1, y1 = prev
            x2, y2 = curr
            if x1 == x2:
                for y in range(min(y1, y2), max(y1, y2) + 1):
                    edge.add((x1, y))
            elif y1 == y2:
                for x in range(min(x1, x2), max(x1, x2) + 1):
                    edge.add((x, y1))
            prev = curr
        return edge

    def check_rectangle(self, a, b):
        x1, y1 = a
        x2, y2 = b
        minX, maxX = min(x1, x2), max(x1, x2)
        minY, maxY = min(y1, y2), max(y1, y2)

        for x in range(minX + 1, maxX):
            if (x, minY + 1) in self.edge or (x, maxY - 1) in self.edge:
                return False
        for y in range(minY + 1, maxY):
            if (minX + 1, y) in self.edge or (maxX - 1, y) in self.edge:
                return False
        return True

    def print_solution(self):
        print("Day 9:")
        print(f"Part 1: {self.part_one()}")
        print(f"Part 2: {self.part_two()}")

